Fix MSW font dialog in high DPI

vadz opened the pull request:
This mostly fixes #22313 but also contains a minor improvement to `AutoSystemDpiAware`.

cc @MaartenBent

vadz committed:
Add an accelerator for the font dialog in the dialogs sample

No real changes, just make it faster to bring up the font dialog for
testing.

vadz committed:
Fix initial point size in MSW wxFontDialog in high DPI

The native dialog doesn't take the actual DPI into account, so convert
our font sizes that do use the DPI to the standard one before using it
and convert font height back to the value corresponding to the actual
DPI once a font is selected in it.

Closes #22313.

vadz committed:
Load SetThreadDpiAwarenessContext() function pointer only once

A small optimization to avoid resolving this function dynamically every
time it is needed and just do it once, on first use.

MaartenBent commented on src/msw/fontdlg.cpp:
```diff
@@ -158,7 +170,24 @@ int wxFontDialog::ShowModal()
+        // being able to determine the correct DPI for this window, due to not
+        // having any valid HWND for it.
+        wxNativeFontInfo nfi(logFont, NULL);
+        nfi.pointSize = wxNativeFontInfo::GetPointSizeAtPPI
+                        (
+                            logFont.lfHeight,
+                            wxDisplay::GetStdPPIValue()
+                        );
```
This is wrong. The point size is already correct, it is `lf.lfHeight` that is wrong (based on 96dpi). You could use `WXAdjustToPPI` to fix it, using the dpi determined before calling `AutoSystemDpiAware`.

I'm not sure if it is needed. When the font is used in wxWidgets it is always adjusted to the DPI, but if a user uses the raw font data with `GetNativeFontInfo()` they might get the wrong `lfHeight`.

vadz replied:
Thanks, this is indeed wrong or, at least, useless, as doing this doesn't change `pointSize` value. I think it's better to use consistent values for `LOGFONT::lfHeight` too, just in case, so I've updated the PR to do it, thanks again for noticing this!

vadz committed:
Change DPI awareness only for ChooseFont() duration

Extract ChooseFont() in a small helper function to ensure that the DPI
awareness is restored as soon as it returns.

No real changes yet.

dconnet commented:
I checked out your branch - it does show the correct pt size now. But now if I ok out and come back in, the font size is now 9 (in my code). In the font dialog, if I use the dialog to select 8pt font, I get 8.5 in the window (I'm not using fractional fonts, so that's probably why I get 9 instead)

vadz commented:
Sorry, I'm a bit lost: what exactly do I need to do in order to see the problem?

It looks like there is some rounding problem here, but I don't see where exactly...

dconnet commented:
In the font sample, it starts as 9pt. I use the Select->Select Font (or ctrl+S) to bring up the font selection (it shows 9). Change the size to 8. Ok. The displayed size is now 8.5. Not all point sizes are bigger. 8, 11, 14, 20, 26 (I stopped there)
This is on a 150% monitor.
On 200%, 8->8.2, 10->9.8, 11->11.2, 14->14.2, 16->15.8, 20->20.2 (stopped there)

vadz committed:
Fix size of the font returned from wxFontDialog in wxMSW

Ensure that the font has exactly the point size entered in the dialog,
rather than having a fractional point size close, but not identical, to
it, as happened before due to first rounding the result of converting
the point size to pixels and then _not_ rounding the result of the
reverse conversion.

This resulted in wxFont::GetPointSize() returning the correct value, but
not GetFractionalPointSize(). Now the latter returns the same, correctly
rounded, value too.

vadz commented:
Thanks, I indeed see it too, but this is not related to high DPI and/or my changes, i.e. I believe this was always the case (or since fractional point size support was added, at least).

The reason is that even in standard DPI (96), there is no integer pixel value corresponding to e.g. 10pt size: the exact value would be `10*96/72 == 13.3(3)` but as `LOGFONT::lfHeight` is an int, it gets rounded down to 13. And, converting back to points, `13*72/96 == 9.75` and not 10.

As the native dialog doesn't allow entering fractional sizes, we can just round the resulting point size to the closest integer and I've done this now.

dconnet commented:
The only issue I'm really seeing now is the sizes that end up at .5. My user has a current font setting of 8. ChooseFont, keep 8. wx returns 8.5, that is being rounded to 9. With dpiAware false or True/PM (and on the 200% monitor), the font stays at 8. But on a 150% screen, I end up in that 8.5 round-trip number.

vadz commented:
Even after the last commit? It should really be always integer now.

dconnet commented:
Yes (confirmed by looking at src after pull).
After selecting 8 in the dialog, the following code reports 9.
```
		auto font = dlg.GetFontData().GetChosenFont();
		wxLogDebug(L"PointSize %d", font.GetPointSize());

```
(just to make sure, I did a full compile on wx and my pgm before this post)

dconnet commented:
Note: You need to be at 150% dpi to see this. 100% or 200% will report back 8.

vadz commented:
> Note: You need to be at 150% dpi to see this. 100% or 200% will report back 8.

I stupidly forgot about this part, sorry. I do see the problem with 150% DPI, thanks.

vadz commented:
I think I've really fixed this now by using the point size to determine the pixel height instead of scaling the latter directly, which could result in an accumulation of errors due to rounding.

Please let me know if anybody still sees anything wrong here.

Thanks again for testing!

dconnet commented:
Much nicer! Thanks! (round tripping now stays consistent)

vadz commented:
Thanks again for testing!
